### ISSUE TITLE
Win32, undecorated, parented GL windows are given the WS_POPUP style

### ISSUE THREAD

**Issue by tonys1110** [panda3d/panda3d#915]
In my project, I am embedding panda3d as a child window inside a wxPython app. I didn't have any luck getting the base classes in direct.wxwidgets to work, so I wrote my own class to bring panda3d and wxPython together.

I got it all working, but noticed that my panda3d window -- a child window inside of a wxFrame -- was still sort of behaving like a top-level frame window despite being a child window. Whenever the panda3d window would receive focus, the main frame window title bar would dim like it was inactive. This would also cause some other minor problems that I can't remember at the moment.

Using Visual Studio's Spy++ tool I realized that the panda3d window has the WS_POPUP style instead of the WS_CHILD style and this was causing the problem. This is despite me doing windowProperties.setUndecorated(True) and windowProperties.setParentWindow(handleToMyWxParentWindow).

I ended up using the separate pywin32 module to hack around this and replace this, but figured I should report this in case anyone else is having the same problem. I also noticed that when the panda3d window is first created it does have WS_CHILD, but then at some point, after creation, WS_CHILD is replaced with WS_POPUP, so I had to add a panda3d Task that watches for this style change and fixes it when it occurs.

```
from direct.showbase.ShowBase import ShowBase
import panda3d.core as pc
import direct.task as dtask

import wx, logging

import win32gui as gui ## pywin32
import win32con as con

log = logging.getLogger(__name__)

class WxPandaWidget(ShowBase):
	def __init__(self, parent):
		ShowBase.__init__(self, fStartDirect=True, windowType='none')

		self.parent = parent
		self.wxApp = wx.App.Get()

		parent.Bind(wx.EVT_SIZE, self.onSize)

		## defer Panda window creation; this is the only way I could get
		## it to initialize/size correctly... this is a one-time task.
		self.startWx()
		self.taskMgr.add(self.onInitialUpdate)


	def onInitialUpdate(self, task):
		"""
		called once to create the Panda window, after the
		parent wxWindow is created and fully initialized.
		"""
		log.info('onInitialUpdate')
		windowProps = pc.WindowProperties.getDefault()
		windowProps.setParentWindow(pc.NativeWindowHandle.makeInt(self.parent.GetHandle()))
		windowProps.setUndecorated(True)
		windowProps.setOrigin(0, 0)
		windowProps.setSize(*self.parent.GetClientSize())
		self.win = self.openWindow(props=windowProps, type='onscreen', unexposedDraw=False)
		self.makeCamera2d(self.win)
		self.setupMouse(self.win)
		self.hwnd = self.win.window_handle.getIntHandle()
		self.taskMgr.add(self.win32StyleWatcher)


	def win32StyleWatcher(self, task):
		"""
		hax to fix window styles. at some point (after creation) panda3d
		gives the GL window the WS_POPUP style instead of WS_CHILD. this
		makes the GL window behave sorta like a top-level window.
		"""
		# hwnd = self.win.window_handle.getIntHandle()
		hwnd = self.hwnd
		curr = prev = gui.GetWindowLong(hwnd, con.GWL_STYLE)

		if (curr & con.WS_POPUP) != 0:
			log.info('trying win32 hax: %d %d %d', self.parent.GetHandle(), hwnd, gui.GetParent(hwnd))
			log.info('  current window style %08x', curr)
			mask = 1
			while mask <= 0x80000000:
				if curr & mask != 0:
					log.info('    %08x', mask)
				mask <<= 1
			curr = (prev & ~con.WS_POPUP) | con.WS_CHILD
			gui.SetWindowLong(hwnd, con.GWL_STYLE, curr)
			gui.SetWindowPos(hwnd, None, 0, 0, 0, 0, con.SWP_FRAMECHANGED |
				con.SWP_NOMOVE | con.SWP_NOSIZE | con.SWP_NOACTIVATE | con.SWP_NOSENDCHANGING | con.SWP_NOZORDER)
			log.info('  done w32 hax %08x -> %08x/%08x', prev, curr, gui.GetWindowLong(hwnd, con.GWL_STYLE))

		return dtask.Task.cont


	def onSize(self, evt):
		windowProps = pc.WindowProperties()
		windowProps.setOrigin(0, 0)
		windowProps.setSize(*self.parent.GetClientSize())
		self.win.requestProperties(windowProps)
		evt.Skip()
```

**Comment by rdb** [comment]
Is this still an issue if you *don't* set the undecorated flag?

What information about properties do you see when setting `notify-level info`?

**Comment by tonys1110** [comment]
Yes this happens if I don't set undecorated (True, False, not calling it at all). notify-level is info, output:

```
:display: loading display module: libpandagl.dll
Known pipe types:
  wglGraphicsPipe
(all display modules loaded.)
:display:windisplay: OS version: 10.0.2.18362
:display:windisplay:   
:display:windisplay: max Mhz 2801000000, current Mhz 2400000000
:ShowBase: Default graphics pipe is wglGraphicsPipe (OpenGL).
:display:windisplay: Got parent_window (7865996)
:display:windisplay: os_handle type NativeWindowHandle::IntHandle
:display:windisplay(warning): SetForegroundWindow() failed!
:display: Created output of type wglGraphicsWindow
:ShowBase: Successfully opened window of type wglGraphicsWindow (OpenGL)
...
:display: Unable to set window properties: parent:(7865996) 
```

That last line happens "long" after initialization is complete and doesn't appear to be in response to any calls that my code is directly making.

**Comment by syntapy** [comment]
I'd like to take a stab at fixing this

Any quick tips on where to start looking?

**Comment by rdb** [comment]
Thanks!  You'd need to look in `panda/src/windisplay/winGraphicsWindow.cxx`, probably in particular the `make_style()` method.

**Comment by syntapy** [comment]
Well I just about gave up on trying to figure out how to get the MSVC debugger to work in C++ code on windows with the python calling app....sigh

But while poking around on the `v1.10.5` tag...I did find out this chunk of code that was added in since `v1.9.4`

https://github.com/panda3d/panda3d/blob/master/panda/src/windisplay/winGraphicsWindow.cxx#L292-L332

I have no idea what that code does but if I just get rid of it, the issue is completely gone for me, it doesnt seem to break my app in any way

**Comment by rdb** [comment]
I wanted to work on this, but can't actually run your code sample.  It's missing the code for providing the `parent` window.  I tried to just create a new wx.Frame and passing that in, but then I can't actually reproduce the issue; it seems to work fine.  Could you post the missing code?

**Comment by rdb** [comment]
Reproduced and fix checked in to 1.10.7 branch.
